### PR TITLE
feat(compiler-cli): DI mode for ctors with default values

### LINKED ISSUE #47606: Allow `inject()` usage as constructor parameter properties default
### Which @angular/* package(s) are relevant/related to the feature request?

compiler-cli, compiler

## Transformer in the way

Injectables' dependencies are currently guessed by analyzing the constructor parameter types.
This typing information is lost after transforming TypeScript to ECMAScript. In order to fix this, the compiler has to run a transformer that converts this information into something like:
```ts
MyCmp.ctorParameters = () => [
   { type: MyService }
];
```
Cf. https://github.com/angular/angular/blob/3a60063a54d850c50ce962a8a39ce01cfee71398/packages/compiler-cli/src/transformers/downlevel_decorators_transform/downlevel_decorators_transform.ts

The problem with this approach is that it slows down community initiatives of integrating Angular with other tools. e.g. Vite, Vitest, SWC...
This forces tools to hack their way through. For instance, jest-preset-angular downloads the internal CLI transformer: https://github.com/thymikee/jest-preset-angular/blob/f24156ea1f7532c02c44d8d023b54623a4bccccd/scripts/prebuild.js#L9

## `inject()` to the rescue

The good news is that an app using only `inject()` instead of constructor param types to retrieve dependencies will not need this transformer:
```ts
class MyCmp {
 myService = inject(MyService);
}
```

The problem with using `inject()` function this way is that it doesn't allow injecting dependencies manually. This means that all tests will need the `TestBed` or similar to control dependencies resolution.
To allow manual instantiation, the current alternative would be something **non-explicit, verbose and error-prone** like this:

```ts
class MyCmp {
   // let's not do this
  constructor(myService?: MyService) {
    this.myService = myService ?? inject(MyService);
  }
}
```

### Proposed solution

Compiler should bypass type-based dependency injection when a default value is set and the given type can't be resolved.
This would allow the following syntax which currently works if this transformer is disabled:

```ts
class MyCmp {
  constructor(private myService = inject(MyService)) {}
}
```

This has the following advantages:
- ✨ constructor parameters transformer is not needed anymore... so easier integration with existing tools
- ✨ manual instantiation is still possible
- ✨ all the benefits of `inject`: type-inference, consistent way of injecting class-based & injection-token based deps: `locale = inject(LOCALE_ID)`, controlling options (e.g. optional, self) through flags instead of decorators.

### Alternatives considered

Use `inject()` and never instantiate manually.

### PR DESCRIPTION
**Pull request by alxhub**
Previously, the AOT compiler would attempt to generate DI constructor calls for all constructors, and report errors if explicit type annotations or other problems prevented them from being generated.

This commit adds a new feature which specially handles constructors that have default values specificed for all parameters. When the compiler encounters such a constructor, it now generates a zero-arg constructor call (`new ClassName()`) as the factory function instead of a normal DI call. This means that the default values (the initializers) of all parameters will be used.

With this feature, the constructor form:

```typescript
constructor(foo = inject(FooService), bar = inject(BarService)) {}
```

now works correctly. Such a constructor can use DI in production, and have these default values overridden for tests.

Fixes #47606

### CODE REVIEW

**Comment by JeanMeche** [comment]
I was looking at #49492 and wondered if something had move on this PR. It seems not. Any update Alex ?

File: packages/core/test/render3/ivy/jit_spec.ts [hunk 1]
```diff
@@ -441,6 +440,48 @@ describe('render3 jit', () => {
...
+
+      const injector = Injector.create({providers: [{provide: Service, useClass: Service}]});
+      const instance = injector.get(Service);
+      expect(instance.injected).toBe(dep);
+    });
+
+
+    it('should construct a type with a destructured parameter in the constructor', () => {
+      @Injectable()
+      class Service {
```
Comment (AndrewKushnir) [inline packages/core/test/render3/ivy/jit_spec.ts:472]: Would it make sense to also add a test with class inheritance where:
- a base class has a constructor with optional params (and another tests with non-optional params)
- a child class that inherits the base one has no constructor

File: packages/core/src/di/jit/constructor.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ */
+export function isProbablyFullyOptionalConstructor(type: Type<any>): boolean {
+  const code = type.toString();
+
+  // Only accept ECMAScript classes
+  if (!code.startsWith('class ')) {
+    return false;
+  }
+
+  // Look for the constructor if one exists.
```
Comment (AndrewKushnir) [inline packages/core/src/di/jit/constructor.ts:31]: nit: it'd be great to add a comment to explain why we return `false` in this case.

**Review by AndrewKushnir — APPROVED** [review]
Looks great! ❤️

File: packages/core/test/render3/ivy/jit_spec.ts (hunk 1 above)
Comment (alxhub) [reply packages/core/test/render3/ivy/jit_spec.ts:472]: I don't think so. The code generation involved here doesn't look at inheritance hierarchy - that's resolved at constructor call time.

**Comment by alxhub** [comment]
Caretaker: the lint check has lost its mind. The formatting that it's complaining about is fine. Formatting the file locally doesn't fix the error, and in fact it keeps cycling between the 'bad' formatting and the 'good' formatting (whether or not a comment in the file is indented).

I've verified that the formatting in the PR is correct, and the error should go away upon merge.

**Comment by JoostK** [comment]
> Caretaker: the lint check has lost its mind. The formatting that it's complaining about is fine. Formatting the file locally doesn't fix the error, and in fact it keeps cycling between the 'bad' formatting and the 'good' formatting (whether or not a comment in the file is indented).
> 
> I've verified that the formatting in the PR is correct, and the error should go away upon merge.

@alxhub I have a fix for the formatting issue in 6c124c58ec94d48f17493f8cc2f5b17e6e82b376, pending in #49136. Merging this PR will likely introduce some more conflicts, so let me rebase that one _after_ this PR has merged so we can get that one in next.

**Comment by yjaaidi** [comment]
@alxhub from my recollection, there was a comment from you saying that this won't work in JIT... but it disappeared and from the JIT tests, I guess that it works. Can you confirm that it is fully operational in JIT too?

**Pull request closed without merging**